File: 프로그래머스/4/43236. 징검다리/징검다리.py
def solution(distance, rocks, n):
    rocks.sort()
    if rocks[-1] != distance:
        rocks.append(distance)
    
    left = 1
    right = distance
    answer = right
    
    while left <= right:
        mid = (left + right) // 2
        
        remove_count = 0
        prev = 0
        for rock in rocks:
            if rock - prev < mid: # ★ mid가 현재 최소여서, mid보다 작으면 제거
                remove_count += 1
            else:
                prev = rock
        
        # mid일 때 n개 이하 제거이면 더 늘려봐도 됨
        if remove_count <= n:
            answer = mid
            left = mid + 1
        else:
            # 너무 많이 제거되었으면 좀 줄여야 함
            right = mid - 1
            
    return answer

File: 프로그래머스/4/43236. 징검다리/징검다리.java
import java.util.*;

class Solution {
    public int solution(int distance, int[] rocks, int n) {
        Arrays.sort(rocks);
        List<Integer> list = new ArrayList<>();
        for(int rock : rocks) list.add(rock);
        list.add(distance); // 마지막 도착 지점 추가

        long left = 1;
        long right = distance; 
        long answer = 0;

        while (left <= right) {
            long mid = (left + right) / 2; // 최소 거리 후보
            int removed = 0;
            int prev = 0;

            for (int rock : list) {
                if (rock - prev < mid) {
                    removed++; // 최소 거리 못 만족하니까 바위 제거
                } else {
                    prev = rock;
                }
            }

            if (removed > n) {
                right = mid - 1; // 너무 많이 제거했으니 최소 거리 줄이기
            } else {
                answer = mid; // 가능한 거리 → 더 큰 거리도 시도
                left = mid + 1;
            }
        }

        return (int) answer;
    }
}
